index-scheduler/src/index_mapper.rs

```diff
@@ -171,6 +186,74 @@ impl IndexMapper {
         Ok(self.index_mapping.get(rtxn, name)?.is_some())
     }
 
+    /// Resizes the maximum size of the specified index to the double of its current maximum size.
+    ///
+    /// This operation involves closing the underlying environment and so can take a long time to complete.
+    pub fn resize_index(&self, rtxn: &RoTxn, name: &str) -> Result<()> {
+        // fixme: factor to a function?
+        let uuid = self
+            .index_mapping
+            .get(rtxn, name)?
+            .ok_or_else(|| Error::IndexNotFound(name.to_string()))?;
+
+        // We remove the index from the in-memory index map.
+        let mut lock = self.index_map.write().unwrap();
+        // check the existing value because we don't want to insert a new resize operation in the map if the index
+        // is currently unavailable.
+        let index = match lock.entry(uuid) {
+            Entry::Occupied(mut occupied) => {
+                // exchange the current index status in the entry with a dummy one so we can take ownership of the index
+                // in the "available" case.
+                // We will reset the index_status to an actual value in the match below.
+                let index_status = std::mem::replace(occupied.get_mut(), BeingDeleted);
+                match index_status {
+                    Available(index) => {
+                        // signal that will be sent when the resize operation completes
+                        let resize_operation = Arc::new(SignalEvent::manual(false));
+                        // temporarily block further attempts to get the index
+                        *(occupied.get_mut()) = BeingResized(resize_operation);
+                        index
+                    }
+                    // the index is already being deleted or resized, so let's ignore this new resize request.
+                    entry @ (BeingDeleted | BeingResized(_)) => {
+                        // reset the index_status to what it was
+                        *(occupied.get_mut()) = entry;
+                        return Ok(());
+                    }
+                }
+            }
+            // the index could not be found, so let's ignore this new resize request.
+            Entry::Vacant(_) => return Ok(()),
+        };
```

Comment:
I just find this part a little bit complex, I don't know if we can simplify it 🤔

Reply:
Deletion operations are synchronous and therefore some match cases seems useless / dead code :)